finality-aleph/src/abft/current/performance/scorer.rs

```diff
@@ -0,0 +1,153 @@
+use current_aleph_bft::{NodeCount, NodeMap, Round};
+
+use crate::{abft::current::performance::Batch, aleph_primitives::RawScore, UnverifiedHeader};
+
+/// Scoring ABFT performance based on returned ordered unit batches.
+pub struct Scorer {
+    newest_unit_by: NodeMap<Round>,
+}
+
+impl Scorer {
+    /// Create a new scorer for the provided node count.
+    pub fn new(node_count: NodeCount) -> Self {
+        Scorer {
+            newest_unit_by: NodeMap::with_size(node_count),
+        }
+    }
+
+    /// Add a batch of ordered units and return a score consisting of numbers of rounds a specific
+    /// node is behind.
+    pub fn process_batch<UH: UnverifiedHeader>(&mut self, batch: Batch<UH>) -> RawScore {
+        let max_round = batch.last().expect("batches always contain a head").round;
+        for unit in batch {
+            // Units are always added in order, so any unit created by an honest node
+            // here has a round greater than any that was included earlier.
+            // This is not necessarily true for forkers, but punishing them is fine.
+            self.newest_unit_by.insert(unit.creator, unit.round)
+        }
+        self.newest_unit_by
+            .size()
+            .into_iterator()
```

Comment:
When I first read this code, I read it wrong so that I though "why it iterates over the same data structure `self.newest_unit_by` using consume iterator and in parallel it reads from it (`self.newest_unit_by.get(node_id)`), but then I realized I read it wrong - the consume iterator is made from `NodeCount` struct, not `NodeMap<Round>`.  Maybe we can improve readability here, by e.g. creating a named variable for ` self.newest_unit_by.size().into_iterator()`? Or simple `for` loop over `0..self.newest_unit_by.size()` would suffice here. My point is, for a reader which does not have this code opened in the IDE, so they can't jump into `.into_iterator()`, this would be more readable.

Reply:
Went with the former, otherwise it's annoying when you cannot `collect` in a loop. Unfortunately the interface for `NodeMap` interates only over present items in general, otherwise this could've been simpler. :/

Reply:
yeah, I've just faced the same when trying to divide parents into direct and non-direct ones